### PR TITLE
Rustbuild cleanups/fixes and improvements

### PR DESCRIPTION
**Pull request by Mark-Simulacrum**
Each commit is a standalone change, and can/should be reviewed separately.

This adds two new functionalities:

 - `--target` and `--host` can be passed without changing config.toml, and we'll respect the users' wishes, instead of requiring that all possible targets are passed.
   - Note that this means that `./x.py clean` won't be quite as wide-spread as before, since it limits itself to the configured hosts, not all hosts. This could be considered a feature as well.
 - `ignore-git` field in `config.toml` which tells Rustbuild to not attempt to load git hashes from `.git`.

This is a precursor to eventual further simplification of the configuration system, but I want to get this merged first so that later work can be made in individual PRs.

r? @alexcrichton

### CODE REVIEW

**Comment by alexcrichton** [comment]
@bors: r+

Nice!

**Comment by kennytm** [comment]
`armhf-gnu` failed, the UI run-pass tests cannot be executed. Cross-testing issue. 

```
[00:42:05] ---- [ui] ui/deprecated-macro_escape-inner.rs stdout ----
[00:42:05] 	ui: /checkout/src/test/ui/deprecated-macro_escape-inner.rs
[00:42:05] thread '[ui] ui/deprecated-macro_escape-inner.rs' panicked at 'failed to exec `/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools/x86_64-unknown-linux-gnu/release/remote-test-client`: Error { repr: Os { code: 2, message: "No such file or directory" } }', /checkout/src/libcore/result.rs:860:4
[00:42:05] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[00:42:05] 
[00:42:05] ---- [ui] ui/deprecated-macro_escape.rs stdout ----
[00:42:05] 	ui: /checkout/src/test/ui/deprecated-macro_escape.rs
[00:42:05] thread '[ui] ui/deprecated-macro_escape.rs' panicked at 'failed to exec `/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools/x86_64-unknown-linux-gnu/release/remote-test-client`: Error { repr: Os { code: 2, message: "No such file or directory" } }', /checkout/src/libcore/result.rs:860:4
[00:42:05] 
[00:42:05] ---- [ui] ui/deriving-meta-empty-trait-list.rs stdout ----
[00:42:05] 	ui: /checkout/src/test/ui/deriving-meta-empty-trait-list.rs
[00:42:05] thread '[ui] ui/deriving-meta-empty-trait-list.rs' panicked at 'failed to exec `/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools/x86_64-unknown-linux-gnu/release/remote-test-client`: Error { repr: Os { code: 2, message: "No such file or directory" } }', /checkout/src/libcore/result.rs:860:4
[00:42:05] 
[00:42:05] ---- [ui] ui/enum-size-variance.rs stdout ----
[00:42:05] 	ui: /checkout/src/test/ui/enum-size-variance.rs
[00:42:05] thread '[ui] ui/enum-size-variance.rs' panicked at 'failed to exec `/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools/x86_64-unknown-linux-gnu/release/remote-test-client`: Error { repr: Os { code: 2, message: "No such file or directory" } }', /checkout/src/libcore/result.rs:860:4
[00:42:05] 
[00:42:05] ---- [ui] ui/issue-19100.rs stdout ----
[00:42:05] 	ui: /checkout/src/test/ui/issue-19100.rs
[00:42:05] thread '[ui] ui/issue-19100.rs' panicked at 'failed to exec `/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools/x86_64-unknown-linux-gnu/release/remote-test-client`: Error { repr: Os { code: 2, message: "No such file or directory" } }', /checkout/src/libcore/result.rs:860:4
[00:42:05] 
[00:42:05] ---- [ui] ui/path-lookahead.rs stdout ----
[00:42:05] 	ui: /checkout/src/test/ui/path-lookahead.rs
[00:42:05] thread '[ui] ui/path-lookahead.rs' panicked at 'failed to exec `/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools/x86_64-unknown-linux-gnu/release/remote-test-client`: Error { repr: Os { code: 2, message: "No such file or directory" } }', /checkout/src/libcore/result.rs:860:4
[00:42:05] 
[00:42:05] ---- [ui] ui/test-should-panic-attr.rs stdout ----
[00:42:05] 	ui: /checkout/src/test/ui/test-should-panic-attr.rs
[00:42:05] thread '[ui] ui/test-should-panic-attr.rs' panicked at 'failed to exec `/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools/x86_64-unknown-linux-gnu/release/remote-test-client`: Error { repr: Os { code: 2, message: "No such file or directory" } }', /checkout/src/libcore/result.rs:860:4
[00:42:05] 
[00:42:05] 
[00:42:05] failures:
[00:42:05]     [ui] ui/deprecated-macro_escape-inner.rs
[00:42:05]     [ui] ui/deprecated-macro_escape.rs
[00:42:05]     [ui] ui/deriving-meta-empty-trait-list.rs
[00:42:05]     [ui] ui/enum-size-variance.rs
[00:42:05]     [ui] ui/issue-19100.rs
[00:42:05]     [ui] ui/path-lookahead.rs
[00:42:05]     [ui] ui/test-should-panic-attr.rs
[00:42:05] 
[00:42:05] test result: FAILED. 348 passed; 7 failed; 1 ignored; 0 measured; 0 filtered out
```

**Comment by Mark-Simulacrum** [comment]
Hm, odd. I can't tell why it wouldn't have been available, since it was built as far as I can tell, and for the right target and stage. @alexcrichton Do we copy the `RemoteTestClient` to the qemu emulator usually? In the meantime, as the queue is mostly empty, I'm going to retry this just in case it was a spurious failure, since I don't believe I changed anything in relevant code...

@bors retry

**Comment by kennytm** [comment]
Same error @Mark-Simulacrum.

**Comment by alexcrichton** [comment]
Hm yeah I dunno what's going on :(

Files like `remote-test-client` indeed never go to the emulator, although this error looks like it's coming from the host, not the emulator? The `remote-test-client` binary was compiled, though, so I'm not sure what's up?

**Comment by Mark-Simulacrum** [comment]
I'm wondering if maybe some change I added makes it so that we clear out the stage0 tools directory at some point? I'll try to investigate soon, though not sure when I'll get the chance.

**Comment by Mark-Simulacrum** [comment]
@bors r=alexcrichton

The latest commit (7d2e1a9) has the fix for the problem encountered in CI. I'm not sure why the problem doesn't occur on master, and the fix contained isn't perfect, but I hope it'll be sufficient. See the commit description for more details.

File: src/bootstrap/compile.rs [hunk 1]
```diff
@@ -198,6 +198,12 @@ impl Step for StdLink {
             // for reason why the sanitizers are not built in stage0.
             copy_apple_sanitizer_dylibs(&build.native_dir(target), "osx", &libdir);
         }
+
+        builder.ensure(tool::CleanTools {
```
Comment (kennytm) [inline src/bootstrap/compile.rs]: Failed to bootstrap. Missing `use tool;`?

```rust
   Compiling bootstrap v0.0.0 (file:///C:/projects/rust/src/bootstrap)
error[E0433]: failed to resolve. Use of undeclared type or module `tool`
   --> src\bootstrap\compile.rs:202:24
    |
202 |         builder.ensure(tool::CleanTools {
    |                        ^^^^^^^^^^^^^^^^ Use of undeclared type or module `tool`
error[E0433]: failed to resolve. Use of undeclared type or module `tool`
   --> src\bootstrap\compile.rs:398:24
    |
398 |         builder.ensure(tool::CleanTools {
    |                        ^^^^^^^^^^^^^^^^ Use of undeclared type or module `tool`
error[E0433]: failed to resolve. Use of undeclared type or module `tool`
   --> src\bootstrap\compile.rs:581:24
    |
581 |         builder.ensure(tool::CleanTools {
    |                        ^^^^^^^^^^^^^^^^ Use of undeclared type or module `tool`
```

**Comment by Mark-Simulacrum** [comment]
I should never trust myself when making code changes without actually checking that they compile.

@bors r=alexcrichton

**Comment by Mark-Simulacrum** [comment]
That's an unexpected failure, but could be relevant -- we may be expecting rustdoc output in a different directory from where we're outputting now. Not certain, though, will need to investigate.

**Comment by aidanhs** [comment]
@bors r-

```
[00:01:39]    --> /checkout/src/bootstrap/config.rs:354:32
[00:01:39]     |
[00:01:39] 343 |             flags.host
[00:01:39]     |             ---------- value moved here
[00:01:39] ...
[00:01:39] 354 |         config.run_host_only = flags.host.is_empty() && !flags.target.is_empty();
[00:01:39]     |                                ^^^^^^^^^^ value used here after move
```
(among others)

**Comment by Mark-Simulacrum** [comment]
I thought I tested this, but apparently not.

@bors r=alexcrichton

**Comment by petrochenkov** [comment]
@bors p=1
That new behavior "rebuild everything after a git commit" is outrageous.

**Comment by kennytm** [comment]
@bors retry #41006

```
Testing libstd stage1 (x86_64-pc-windows-gnu -> x86_64-pc-windows-gnu)
   Compiling rand v0.0.0 (file:///C:/projects/rust/src/librand)
   Compiling core v0.0.0 (file:///C:/projects/rust/src/libcore)
   Compiling std_unicode v0.0.0 (file:///C:/projects/rust/src/libstd_unicode)
   Compiling alloc v0.0.0 (file:///C:/projects/rust/src/liballoc)
   Compiling collections v0.0.0 (file:///C:/projects/rust/src/libcollections)
error: failed to remove: C:\projects\rust\build\x86_64-pc-windows-gnu\stage1-std\x86_64-pc-windows-gnu\release\std.dll
Caused by:
  Access is denied. (os error 5)
```

**Comment by kennytm** [comment]
`x86_64-pc-windows-msvc` (check-aux) failed, cargo's [build](https://github.com/rust-lang/cargo/blob/master/tests/build.rs) tests' doc-tests failed due to "crate `std` not found". Seems legit. (Queue is pretty full, not going to retry this time)

```
--- stdout
running 0 tests
test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
--- stderr
   Compiling world v1.0.0 (file:///C:/projects/rust/build/x86_64-pc-windows-msvc/stage2-tools/x86_64-pc-windows-msvc/cit/t73/world)
    Finished dev [unoptimized + debuginfo] target(s) in 0.85 secs
     Running target\debug\deps\world-142ff8362dbaa2cb.exe
   Doc-tests world
error[E0463]: can't find crate for `std`
error: aborting due to previous error
thread '<unnamed>' panicked at 'Box<Any>', src\librustc_errors\lib.rs:524:8
note: Run with `RUST_BACKTRACE=1` for a backtrace.
error: test failed, to rerun pass '--doc'
', C:\Users\appveyor\.cargo\registry\src\github.com-1ecc6299db9ec823\hamcrest-0.1.1\src\core.rs:31:12
failures:
    crate_authors_env_vars
    crate_env_vars
    explicit_examples
    implicit_examples
test result: FAILED. 111 passed; 4 failed; 0 ignored; 0 measured; 0 filtered out
error: test failed, to rerun pass '--test build'
```

**Comment by Mark-Simulacrum** [comment]
I don't have access to a windows machine to attempt reproduction there, but all attempts at reproducing locally (including an aux container build for Travis) haven't succeeded, so I'm going to assume this was a spurious failure somehow. I guess retrying is the best idea I have right now (we've built libstd at this point for sure for all stages and the right targets, I think). If it's not a spurious failure then I guess it probably has to do with building the rustdoc we're using for the Cargo tests with the stage2 compiler (instead of the stage1 compiler, as we used to do before this PR), though I don't know what the problem there could be either.

@bors retry

**Comment by alexcrichton** [comment]
@bors: r-

@Mark-Simulacrum I think it's becuse the `Cargo` check step doesn't depend on rustdoc?

I'm wary to land spurious failures :(

**Comment by Mark-Simulacrum** [comment]
We've built rustdoc by this point, though, and we depend on it from the `cargo` method on `Builder`, so that doesn't seem like a valid reason. Could you trying running `./x.py test src/tools/cargo` on a windows machine if you have a chance with this PR? I would hope to see failures, but if not I'm at a loss.

In general, I'm confused why these particular tests are failing since doc-tests are working, for cargotest (e.g., for iron). My only suspicion is that we're not properly clearing the rustdoc environment variable or something like that, though I can't find a spot where we'd be doing that.

**Comment by alexcrichton** [comment]
I've started a build locally, will get back with results.

**Comment by alexcrichton** [comment]
Ok I have reproduced this locally, I will attempt to dig into this when I can. (failure is the same as CI)

**Comment by alexcrichton** [comment]
Ok here's what's happening:

* At the root, Cargo's tests are running the wrong `rustdoc` executable
  * tests are executing `C:\\msys64\\code\\rust2\\build\\x86_64-pc-windows-msvc\\stage2-tools\\x86_64-pc-windows-msvc\\release\\rustdoc.exe`
  * test should execute `C:\\msys64\\code\\rust2\\build\\x86_64-pc-windows-msvc\\stage2\\bin\\rustdoc.exe`
  * Both of these are in `PATH`, but they're listed in the wrong order
* How did we get here?
  * rustbuild [configures `RUSTDOC`](https://github.com/rust-lang/rust/blob/456493353867832170045cb063aebf22fa8a7201/src/bootstrap/builder.rs#L484)
  * cargo [ignores this](https://github.com/rust-lang/cargo/blob/22e5c7678762c10690540072131155d2b8d95a03/tests/cargotest/lib.rs#L55)
  * Cargo has a [rustdoc process builder](https://github.com/rust-lang/cargo/blob/master/src/cargo/ops/cargo_rustc/compilation.rs#L86-L89) which [modifies the dylib search path](https://github.com/rust-lang/cargo/blob/master/src/cargo/ops/cargo_rustc/compilation.rs#L126-L143), which on windows is `PATH` itself.
  * For "target processes" (which this is, oddly enough) Cargo will add a bunch of paths. For example my test locally adds:

    ```
    C:\msys64\code\rust2\build\x86_64-pc-windows-msvc\stage2-tools\x86_64-pc-windows-msvc\release\build\curl-sys-e1c62a2753fe248f\out/lib
    C:\msys64\code\rust2\build\x86_64-pc-windows-msvc\stage2-tools\x86_64-pc-windows-msvc\release\build\libgit2-sys-430a3893b65ec063\out/lib
    C:\msys64\code\rust2\build\x86_64-pc-windows-msvc\stage2-tools\x86_64-pc-windows-msvc\release\build\libz-sys-e4986560bd812fa7\out/lib
    C:\msys64\code\rust2\build\x86_64-pc-windows-msvc\stage2-tools\x86_64-pc-windows-msvc\release\build\libssh2-sys-1dfafe498be1b09e\out/lib
    C:\msys64\code\rust2\build\x86_64-pc-windows-msvc\stage2-tools\x86_64-pc-windows-msvc\release\build\miniz-sys-8ba7ad34c6f6015f\out
    C:\msys64\code\rust2\build\x86_64-pc-windows-msvc\stage2-tools\x86_64-pc-windows-msvc\release
    C:\msys64\code\rust2\build\x86_64-pc-windows-msvc\stage2-tools\x86_64-pc-windows-msvc\release\deps
    ```

  * As part of this PR rustdoc previously appeared in `stage1-tools` and now it's appearing in `stage2-tools`, namely `build\x86_64-pc-windows-msvc\stage2-tools\x86_64-pc-windows-msvc\release`

So this only happens on Windows b/c the dylib load path is `PATH`, and this only happens on this PR b/c rustdoc is showing up in `stage2-tools`, not `stage1-tools`.

tl;dr; let's just call the binary `rustdoc-some-random-name` and then copy it to the sysroot under the name `rustdoc.exe`, that should fix things.

**Comment by Mark-Simulacrum** [comment]
Thanks for investigating! I've pushed a fix that just renames the "tool" rustdoc to `rustdoc-tool-binary` which should work if I understand correctly, and rebased.

@bors r+

**Comment by Mark-Simulacrum** [comment]
Ah,

@bors r=alexcrichton

**Comment by Mark-Simulacrum** [comment]
Compiler hosts aren't always capable of running on the current platform...

@bors r=alexcrichton
